### ISSUE TITLE
Slight audio skipping + glitches on GoatTracker 2 tunes

### ISSUE THREAD

**Issue by ruby-R53** [libsidplayfp/libsidplayfp#161]
The most notable example I have is Jeroen Tel & LMan's `Skypeople.sid`.

At the very beginning, voices 1 and 2 sometimes don't get in sync, making the chords almost sound like arps. Then, at around 1 minute of playback, I hear some audio samples getting skipped, which not only makes it sound cut-off in parts, it also seems to glitch the ADSR a little bit. The downward arpeggios at around 1:24 sound a bit louder at the beginning and then they quickly go back to their proper volume level.

I also noticed the problem gets a bit worse when my PC is under some load like a compiling process, which is when the latter happens.

If I record that same tune to a .wav however, none of that happens.

**Comment by acrouzet** [comment]
Skypeople uses CheeseCutter, not GoatTracker 2.

**Comment by ruby-R53** [comment]
huh, really? i remember seeing GT being mentioned on that tune in DeepSID, maybe i mixed it with something else

well the other tunes i know that were actually made in GT2 and showed these same issues were ones that i made a few months ago

so i guess there's something in the code generated by both these programs that makes their .sid's sound off in libsidplayfp

**Comment by ruby-R53** [comment]
i just remembered another tune that has the same issue, however it wasn't made in GT: Thomas & Michael Detert's `Gordian_Tomb.sid`

you can notice a few samples getting skipped or getting very slightly stretched (?) at the beginning of it, before voice 1 comes into action

so i better change the title now lol

**Comment by ruby-R53** [comment]
yeah, another tune with the same problem is AMJ's `Pashaus.sid` when voice 1 plays the saw+pulse parts and when voice 3 does that triangle↔noise drum

**Comment by reFX-Mike** [comment]
I found that with a lot of tunes, there are small skips. It's not related to the tunes themselves, but rather libsidplayfp. Sometimes, I can listen to a tune without any problems, and the next time I play the exact same tune, it suddenly has a small skip somewhere.

**Comment by ruby-R53** [comment]
yup exactly what i experience as well

**Comment by reFX-Mike** [comment]
I believe the re-sampling process from 1 MHz to the audio device's sample rate occasionally skips a buffer.

**Comment by ruby-R53** [comment]
yeah that could be the case, which'd explain why it doesn't happen when recording to a file maybe

**Comment by reFX-Mike** [comment]
That would not explain it, as the same function is still used, but maybe with a different sample rate. So you never ever had the same skips when rendering to a file?

**Comment by ruby-R53** [comment]
ahh right, i can hear it now on the WAV renders too but it's far more subtle somehow

**Comment by reFX-Mike** [comment]
Because it depends on the output sample rate. Your realtime audio is probably at 48 kHz, whereas you render might be 44.1 kHz.

**Comment by ruby-R53** [comment]
nope they're both at 192 kHz here

weird

**Comment by reFX-Mike** [comment]
No idea what happened, but I haven’t had these glitches in a while now.

**Comment by ruby-R53** [comment]
i noticed these glitches are more pronounced on my laptop which has a weaker hardware, a while back i upgraded my pc cpu and i didn't notice them

guess the solution is to get a better hardware ¯\_(ツ)_/¯

**Comment by reFX-Mike** [comment]
Nah, that's not it. That wouldn't explain why even rendering into a WAV used to skip here and there. I frequently run debug versions that are 10x slower than release builds, and it still doesn't skip anymore. It's a mystery now.

**Comment by ruby-R53** [comment]
UPDATE: i'm getting `Segmentation fault` when trying to play with both PulseAudio and ALSA, and with out123 i don't get any sound at all and the display updates way too fast, then when i quit with ESC i get

```
double free or corruption (!prev)
Aborted
```

i can still render the .sid to a file tho', and even then i still notice the same audio glitches, so that's weird

**Comment by ruby-R53** [comment]
UPDATE (again):
> i noticed these glitches are more pronounced on my laptop which has a weaker hardware

welp i just tried playing `Skypeople.sid` on FreeBSD (which's installed on my PC along with Gentoo) and it's even worse on it for some reason 🫠

**Comment by mywave82** [comment]
Does it happen if you use alsa without it going via pulseaudio (default typically is with libpulse, so need to specify hw directly.

(Reminds me about some random behaviour I have seen others experience with pulseaudio in other audio software)

**Comment by drfiemost** [comment]
> umm i'm getting an error again
> 
> ![Image](https://github.com/user-attachments/assets/98080600-f080-4efd-83f0-024e5544ce60)

This is due to missing -lpthread in the pkg-config file, pleas open a new issue

**Comment by ruby-R53** [comment]
nah already fixed it, just added that line x)

> Does it happen if you use alsa without it going via pulseaudio (default typically is with libpulse, so need to specify hw directly.

yup it still does somehow

**Comment by drfiemost** [comment]
Does FreeBSD come with pulseaudio? Or it just uses OSS?

**Comment by ruby-R53** [comment]
i'm not really sure, it doesn't come with pulseaudio but it does have sndio and oss although i dunno which of those i'm running at the moment

**Comment by ruby-R53** [comment]
hey, i just played these same tunes after the new player API implementation and they got a lot more subtle now, even for `Pashaus.sid`

for `Skypeople.sid` the bug seems to have been completely fixed, same thing for `Ice_Age.sid` too

however the only excpetion is `Cold_War_Generation.sid` where i still hear the exact same glitches

so i guess it's rather a positive side effect of that implementation and not that it actually fixed it
